src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -1337,10 +1325,180 @@ void ShenandoahHeap::object_iterate(ObjectClosure* cl) {
+    _bitmap(bitmap), _queue(q), _heap(ShenandoahHeap::heap()),
+    _marking_context(_heap->marking_context()) {}
+  void do_oop(oop* p)       { do_oop_work(p); }
+  void do_oop(narrowOop* p) { do_oop_work(p); }
+};
+
+// Object iterator for parallel heap iteraion.
+// The root scanning phase happenes in construction as a preparation of
+// parallel marking queues.
+// Every worker processes it's own marking queue. work-stealing is used
+// to balance workload.
+class ShenandoahParallelObjectIterator : public ParallelObjectIterator {
+private:
+  uint                         _num_workers;
+  bool                         _init_ready;
+  MarkBitMap*                  _aux_bit_map;
+  ShenandoahHeap*              _heap;
+  Stack<oop, mtGC>             _roots_stack; // global roots stack
+  ShenandoahObjToScanQueueSet* _task_queues;
+public:
+  ShenandoahParallelObjectIterator(uint num_workers, MarkBitMap* bitmap) :
+        _num_workers(num_workers),
+        _init_ready(false),
+        _aux_bit_map(bitmap),
+        _heap(ShenandoahHeap::heap()) {
+    // Initialize bitmap
+    _init_ready = _heap->prepare_aux_bitmap_for_iteration();
+    if (!_init_ready) {
+      return;
+    }
+    // root marking
+    ObjectIterateScanRootClosure oops(_aux_bit_map, &_roots_stack);
+    // Procssing roots
+    _heap->scan_roots_for_iteration(&_roots_stack, &oops);
+    // prepare worker stacks
+    _init_ready = prepare_worker_queues();
+  }
+
+  ~ShenandoahParallelObjectIterator() {
+    // Reclaim bitmap
+    _heap->reclaim_aux_bitmap_for_iteration();
+    // Reclaim queue for workers
+    if (_task_queues!= NULL) {
+      for (uint i = 0; i < _num_workers; ++i) {
+        ShenandoahObjToScanQueue* q = _task_queues->queue(i);
+        if (q != NULL) {
+          delete q;
+          _task_queues->register_queue(i, NULL);
+        }
+      }
+      delete _task_queues;
+      _task_queues = NULL;
+    }
   }
+
+  virtual void object_iterate(ObjectClosure* cl, uint worker_id) {
+    if (_init_ready) {
+      object_iterate_parallel(cl, worker_id, _task_queues);
+    }
+  }
+
+private:
+  // divide global root_stack into worker queues.
+  bool prepare_worker_queues() {
+    _task_queues = new ShenandoahObjToScanQueueSet((int) _num_workers);
+    // initialize queue for every workers
+    for (uint i = 0; i < _num_workers; ++i) {
+      ShenandoahObjToScanQueue* task_queue = new ShenandoahObjToScanQueue();
+      task_queue->initialize();
+      _task_queues->register_queue(i, task_queue);
+    }
+    // divide roots to every worker
+    size_t roots_num = _roots_stack.size();
+    if (roots_num == 0) {
+      // no work to do.
+      return false;
+    }
+    // assume that object referencing distribution is related with
+    // root kind, use round-robin to make every worker have same chance
+    // to process every kind of roots
+    for (uint j = 0; j < roots_num; j++) {
+      uint stack_id = j % _num_workers;
+      oop obj = _roots_stack.pop();
```

Comment:
I am not sure we can `pop()` the stack from multiple threads without any sort of synchronization. It might be better to just walk the same roots from multiple threads, and then let the bitmap help us figure out if we have visited the root already. See how `ShenandoahVerifierReachableTask` does it?

Reply:
Hi @shipilev, 
Thanks for your effort for reviewing this change!
The method prepare_worker_queue() here is invoked in the constructor of ShenandoahParallelObjectIterator, which is designed to run in single thread.  The logic of using ParallelObjectIterator is as follows: (refer to code in hotspot/share/memory/heapInspection.cpp)
 -  create instance of ParallelObjectIterator in shenandoahHeap:
      ParallelObjectIterator* poi = Universe::heap()->parallel_object_iterator(gang->active_workers())
 - create heap iteration task and run (invoke ParallelObjectIterator's method object_iterator())
     ParHeapInspectTask task(**poi**, cit, filter);
     // Run task with the active workers.
     gang->run_task(&task); 
=======
So the ShenandoahParallelObjectIterator iterate all roots in the single thread, collecting all roots in the roots_stack, then prepare the queues for every workers.  -- processed in constructor.
And the workers start their processing from their queue, and then iterate all reachable objects. -- processed in work(uint worker_id)

I have checked that ShenandoahVerifierReachableTask supports iterating roots parallelly, I didn't choose that way because I want to make every workers to start with nearly the same number of roots in their queue, hope it could help on inital work load balance.

Lin

Reply:
Ah! I missed that part. I thought stack is used by the actual parallel workers. Nevermind my correctness comment then.